src/components/scenes/ChangePasswordScene.tsx

```diff
@@ -79,48 +83,49 @@ const ChangePasswordSceneComponent = ({
     .some(([, value]) => value === 'unmet' || value === 'error')
 
   const handleNext = useHandler(async () => {
-    if (password !== confirmPassword) {
-      setConfirmPassword('')
-      setPasswordReqs({
-        ...passwordReqs,
-        confirmationMatches: 'error'
-      })
-      return
-    }
-
     setSpinning(true)
 
-    try {
-      await onSubmit(password)
-    } catch (e) {
-      showError(e)
-      setSpinning(false)
+    handleValidatePassword(password, 'error')
+    if (
+      !Object.entries(passwordReqs).some(
+        ([, value]) => value === 'unmet' || value === 'error'
+      )
+    ) {
+      try {
+        await onSubmit(password)
+      } catch (e) {
+        showError(e)
+      }
     }
+
+    setSpinning(false)
   })
 
   const handleFocusSwitch = () => {
     setFocusFirst(false)
     setFocusSecond(true)
   }
 
-  const handleValidatePassword = (password: string) => {
-    setPassword(password)
-
-    setPasswordReqs({
-      minLengthMet: password.length >= 10 ? 'met' : 'unmet',
-      hasNumber: /[0-9]/.test(password) ? 'met' : 'unmet',
-      hasLowercase: /[a-z]/.test(password) ? 'met' : 'unmet',
-      hasUppercase: /[A-Z]/.test(password) ? 'met' : 'unmet'
-    })
-  }
+  const handleValidatePassword = useHandler(
+    (password: string, failStatus: PasswordRequirementStatus) => {
+      setPasswordReqs({
```

Comment:
Calling `setPasswordReqs` will only take effect on the next render. Until that happens, `passwordReqs` will keep whatever value it had a the previous render. This is just how the `useState` hook operates.

This means that the `handleNext` function will be looking at outdated password requirements, wihich seems like a problem.

One way to resolve this is to not even bother calling `validatePassword` from `handleNext`. Or, if you really do want the fresh results, you should have `validatePassword` return the newly-calculated requirements object back out to `handleNext`, so it can check the fresh version.

Reply:
That makes sense, though I can't seem to replicate this kind of a bug when
actually testing... I'll make the change regardless, though.